proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java

```diff
@@ -652,28 +652,42 @@ private void finish2()
         }
 
         ch.getHandle().pipeline().get( HandlerBoss.class ).setHandler( new UpstreamBridge( bungee, userCon ) );
-        bungee.getPluginManager().callEvent( new PostLoginEvent( userCon ) );
 
-        // #3612: Don't progress further if disconnected during event
-        if ( ch.isClosed() )
+        Callback<PostLoginEvent> complete = new Callback<PostLoginEvent>()
         {
-            return;
-        }
+            @Override
+            public void done(PostLoginEvent result, Throwable error)
+            {
+                // #3612: Don't progress further if disconnected during event
+                if ( ch.isClosed() )
+                {
+                    return;
+                }
 
-        ServerInfo server;
-        if ( bungee.getReconnectHandler() != null )
-        {
-            server = bungee.getReconnectHandler().getServer( userCon );
-        } else
-        {
-            server = AbstractReconnectHandler.getForcedHost( InitialHandler.this );
-        }
-        if ( server == null )
-        {
-            server = bungee.getServerInfo( listener.getDefaultServer() );
-        }
+                ServerInfo server;
+                if ( bungee.getReconnectHandler() != null )
+                {
+                    server = bungee.getReconnectHandler().getServer( userCon );
+                } else
+                {
+                    server = AbstractReconnectHandler.getForcedHost( InitialHandler.this );
+                }
+                if ( server == null )
+                {
+                    server = bungee.getServerInfo( listener.getDefaultServer() );
+                }
```

Comment:
Move this part to before calling the event and expose the initial target server by the reconnect handler to plugins?

Reply:
Added the target server before the event and added a value in the constructor to show the initial target server to developers.